BCC CyHy reports and CyHy notifications to the appropriate CSA

jsf9k opened the pull request:
## 🗣 Description ##

This pull request makes the necessary changes to BCC CyHy reports and CyHy notifications to the appropriate CISA Cyber Security Advisor (CSA).

## 💭 Motivation and context ##

Partly resolves cisagov/cyhy-system#114.  See also cisagov/cyhy_amis#749.

## 🧪 Testing ##

All automated tests pass.  I did a full dry run (i.e., did everything except actually tell AWS SES to send the email) of sending the CyHy reports and everything worked as expected.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Finalize version.

## ✅ Post-merge checklist ##

- [x] Add a tag (but don't create a release).
- [x] Create a local Docker image: `docker build -t cisagov/cyhy-mailer:1.6.0 --platform=linux/amd64 .`
- [x] Push Docker image up to Docker Hub: `docker push cisagov/cyhy-mailer:1.6.0`
- [x] Deploy this change to Production.

jsf9k committed:
Copy in the latest pre-commit config from cisagov/skeleton-python-library

This is necessary because pre-commit does not work with the most
recent version of Python with the old config, ans this repository is
not skeletonized.

jsf9k committed:
Add an __init__.py file

The mypy linter requires this fine be present in order to pass.

jsf9k committed:
Drop explicit "r" arguments when calling open()

This is required by the pyupgrade linter.

jsf9k committed:
Grab the state abbreviation when pulling agency data from the DB

We will need this information to lookup the appropriate CISA Cyber
Security Advisor's email address.

jsf9k committed:
Add CLI code to automatically BCC appropriate CSAs on all CyHy emails

These emails include CyHy reports and CyHy notifications.

jsf9k committed:
Add secret and CLI option for CSA emails to Docker compositions

Thus far only the CyHy reports and CyHy notifications have been
specified as needing to be sent along to the CSAs.

jsf9k committed:
Bump version from 1.5.2 to 1.6.0

jsf9k committed:
Bump version from 1.6.0 to 1.6.0-rc.1

jsf9k committed:
Add error handling for case where agency lacks state information

Also add some debug-level logging to show what CSA addresses are being
BCCd.

jsf9k committed:
Handle error case where agency's state is not associated with a CSA region

jsf9k committed:
Update README to include new CLI argument

dav3r commented on cyhy/mailer/cli.py:
```diff
@@ -1060,6 +1145,32 @@ def main():
+    # Rearrange the CSA email mapping data into a more convenient
+    # form.
+    csa_emails = {}
+    for region in csa_email_mapping:
```
If `csa_email_file` is `None`, won't `csa_email_mapping` will be undefined here?

(I agree with you that the `--csa-email-file` flag should be optional; if it is not provided, then we should just use the default BCC list.)

jsf9k replied:
Please see commit [8dda8e5](https://github.com/cisagov/cyhy-mailer/pull/101/commits/8dda8e59ba92bc2df5b9e5d99babf14478e62539).

dav3r replied:
I don't think I fully got my point across in my last comment.  I was trying to say that it should be possible for someone to use cyhy-mailer to send out CyHy reports and notifications, but not be forced to use the CSA BCC functionality being added in this PR.  I know this was not explicitly stated in https://github.com/cisagov/cyhy-system/issues/114, but I don't see any reason why this couldn't be an optional feature.  If the `--csa-email-file` isn't provided, the code should not attempt to look up CSA emails at all; it should behave as it does now and use the default BCC list.

Please let me know if this makes sense or not.

jsf9k replied:
Please see commit [ffccc70](https://github.com/cisagov/cyhy-mailer/pull/101/commits/ffccc70d93e614d8e4dbe5eb7ff8ae77fce8acfe).  I tested by doing:
- A full CyHy reports dry run where the CSA email YAML was specified
- A full CyHy reports dry run without specifying the CSA email YAML

dav3r replied:
Yes, that is the cacafuego that I'm talking about! 🌶️ 💩 🔥 
Thanks for making it happen.

dav3r commented on cyhy/mailer/cli.py:
```diff
@@ -1060,6 +1145,32 @@ def main():
 
+    # Parse out the CSA email mapping, if present.
+    csa_email_file = args["--csa-email-file"]
+    if csa_email_file is not None:
```
Can `csa_email_file` ever be `None` since we are setting a default value for it in the docopt options above(`/run/secrets/csa_emails.yml`)?
https://github.com/cisagov/cyhy-mailer/blob/8aecdbfef831a091aab86a5c189d7e5a8aef9a5d/cyhy/mailer/cli.py#L34-L50

jsf9k replied:
Please see commit [8dda8e5](https://github.com/cisagov/cyhy-mailer/pull/101/commits/8dda8e59ba92bc2df5b9e5d99babf14478e62539).

jsf9k replied:
Please see commit [ffccc70](https://github.com/cisagov/cyhy-mailer/pull/101/commits/ffccc70d93e614d8e4dbe5eb7ff8ae77fce8acfe).  I tested by doing:
- A full CyHy reports dry run where the CSA email YAML was specified
- A full CyHy reports dry run without specifying the CSA email YAML

dav3r replied:
🌶️ 💩 🔥

dav3r reviewed (commented):
I have 2 questions that contradict one another, but I'm pretty sure one of them is valid. 😆

jsf9k committed:
Make the CSA email file argument optional

The argument is only necessary if --cyhy-report-dir or
--cyhy-notification-dir is present.

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Make the CSA email file argument optional, even for CyHy reports and notifications

Now CyHy reports and notifications will simply not be BCCd to the CSAs
if the argument is not present.

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

dav3r commented on cyhy/mailer/cli.py:
```diff
@@ -761,6 +793,27 @@ def send_cyhy_reports(db, batch_size, ses_client, cyhy_report_dir, dry_run=False
+                    state = request["agency"]["location"]["state"]
+                except KeyError:
+                    logging.warn(
+                        f"State or territory information unavailable for agency with ID {id}.  As a result this CyHy report cannot be sent to the corresponding CSA."
+                    )
+                else:
+                    try:
+                        csa_email_address = csa_emails[state]
+                    except KeyError:
+                        logging.warn(
+                            f"State or territory {state} associated with agency with ID {id} does not correspond to a CSA region.  As a result this CyHy report cannot be sent to the corresponding CSA."
+                        )
+                    else:
+                        logging.debug(
+                            f"BCCing report for agency with ID {id} to {csa_email_address} because the agency is associated with the state or territory {state}."
```
Even though our request document structure still calls everyone an "agency", I think at this point, we should be shifting to either "entity" or "organization" (my preference is "entity"):
```suggestion
                        f"State or territory information unavailable for entity with ID {id}.  As a result this CyHy report cannot be sent to the corresponding CSA."
                    )
                else:
                    try:
                        csa_email_address = csa_emails[state]
                    except KeyError:
                        logging.warn(
                            f"State or territory {state} associated with entity with ID {id} does not correspond to a CSA region.  As a result this CyHy report cannot be sent to the corresponding CSA."
                        )
                    else:
                        logging.debug(
                            f"BCCing report for entity with ID {id} to {csa_email_address} because the entity is associated with the state or territory {state}."
```

Though it might make more sense to log an issue and take care of this wholesale in this repo?

jsf9k replied:
Please see commits 0e0231f and 840fd56.

jsf9k replied:
This request caused me to notice [another issue](https://github.com/cisagov/cyhy-mailer/issues/102), which I have assigned to you.

dav3r replied:
Lovely! 👍 👍

dav3r commented on cyhy/mailer/cli.py:
```diff
@@ -949,6 +1015,27 @@ def send_cyhy_notifications(
+                state = request["agency"]["location"]["state"]
+            except KeyError:
+                logging.warn(
+                    f"State or territory information unavailable for agency with ID {id}.  As a result this CyHy report cannot be sent to the corresponding CSA."
+                )
+            else:
+                try:
+                    csa_email_address = csa_emails[state]
+                except KeyError:
+                    logging.warn(
+                        f"State or territory {state} associated with agency with ID {id} does not correspond to a CSA region.  As a result this CyHy report cannot be sent to the corresponding CSA."
+                    )
+                else:
+                    logging.debug(
+                        f"BCCing report for agency with ID {id} to {csa_email_address} because the agency is associated with the state or territory {state}."
```
Same as previous comment:
```suggestion
                    f"State or territory information unavailable for entity with ID {id}.  As a result this CyHy report cannot be sent to the corresponding CSA."
                )
            else:
                try:
                    csa_email_address = csa_emails[state]
                except KeyError:
                    logging.warn(
                        f"State or territory {state} associated with entity with ID {id} does not correspond to a CSA region.  As a result this CyHy report cannot be sent to the corresponding CSA."
                    )
                else:
                    logging.debug(
                        f"BCCing report for entity with ID {id} to {csa_email_address} because the entity is associated with the state or territory {state}."
```

jsf9k replied:
Please see commits 0e0231f and 840fd56.

dav3r reviewed (commented):
One more thang that caught my eye:

jsf9k committed:
Prefer "entity" to "agency" in all comments and print/logging statements

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Prefer "entity" to "agency" in variable names

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Replace deprecated function from datetime library

datetime.datetime.utcnow() is now deprecated and
datetime.datetime.now(datetime.UTC) is preferred.

jsf9k committed:
Change pre-formatted text styles from bash to console

There is no bash code in these pre-formatted text blocks.

dav3r reviewed (approved):
Looks great, thanks for making those changes! 🚀 

I'll take care of fixing those failing tests (#102).

dav3r committed:
Merge branch 'develop' into feature/email-csas

mcdonnnj commented on .pre-commit-config.yaml:
```diff
@@ -81,48 +106,65 @@ repos:
+      - id: mypy
+        additional_dependencies:
+          - types-PyYAML
+          - types-setuptools
```
This adds typing for other packages used and adjusts the arguments being used by the hook.

```suggestion
        additional_dependencies:
          - boto3-stubs
          - pymongo-stubs
          - types-chevron
          - types-docopt
          - types-PyYAML
          - types-setuptools
        args:
          # Default arguments
          - --ignore-missing-imports
          - --scripts-are-modules
          # Support native namespace packages with pre-commit
          - --explicit-package-bases
```

mcdonnnj commented on cyhy/__init__.py:
Please delete this file along with the pre-commit configuration suggestion. The layout in this project uses [native namespace packages](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages) and so I do not think this file should be present.

dav3r replied:
👍 👍   Please see https://github.com/cisagov/cyhy-mailer/pull/101/commits/9f1e7927e7602493ce3da689bb252a2f11753703.

mcdonnnj commented on cyhy/mailer/cli.py:
```diff
@@ -31,6 +31,27 @@
   -c --db-creds-file=FILENAME       A YAML file containing the Cyber
                                     Hygiene database credentials.
                                     [default: /run/secrets/database_creds.yml]
+  -e --csa-email-file=FILENAME      A YAML file associating each state
```
Since this file should associate multiple emails I think the argument should use the plural. Note: I spaced and accompanying suggestions are further down.

```suggestion
  -e --csa-emails-file=FILENAME     A YAML file associating each state
```

mcdonnnj commented on docker-compose.cyhy-notification.yml:
```diff
@@ -11,4 +18,5 @@ services:
       - notification
       - --cyhy-notification-dir=/cyhy_notifications/
       - --db-creds-file=/run/secrets/database_creds.yml
+      - --csa-email-file=/run/secrets/csa_emails.yml
```
Matches my suggestion for the cli arguments.

```suggestion
      - --csa-emails-file=/run/secrets/csa_emails.yml
```

mcdonnnj commented on docker-compose.cyhy.yml:
```diff
@@ -16,4 +23,5 @@ services:
       - --cyhy-report-dir=/cyhy_reports/
       - --cybex-scorecard-dir=/cybex_scorecard/
       - --db-creds-file=/run/secrets/database_creds.yml
+      - --csa-email-file=/run/secrets/csa_emails.yml
```
Matches my suggestion for the cli arguments.

```suggestion
      - --csa-emails-file=/run/secrets/csa_emails.yml
```

mcdonnnj commented on cyhy/mailer/cli.py:
```diff
@@ -203,6 +225,7 @@ def get_requests_raw(db, query, batch_size=None):
         "agency.contacts.name": True,
         "agency.contacts.email": True,
```
Alphabetization since we're here.

```suggestion
        "agency.contacts.email": True,
        "agency.contacts.name": True,
```

mcdonnnj commented on README.md:
```diff
@@ -56,6 +55,26 @@ Options:
                                     [default: /run/secrets/database_creds.yml]
+  -e --csa-email-file=FILENAME      A YAML file associating each state or
+                                    territory with the email address of the
+                                    corresponding CISA Cyber Security Advisor
```
One word!

```suggestion
                                    corresponding CISA Cybersecurity Advisor
```

dav3r replied:
Per https://www.cisa.gov/about/regions/security-advisors, we actually do refer to CSAs as "Cyber Security Advisors" (sigh).

mcdonnnj replied:
See in
- https://www.cisa.gov/region-3-what-we-do
- https://www.cisa.gov/region-4-what-we-do
- https://www.cisa.gov/region-5-what-we-do
- etc. (every region I manually checked)

they use "Cybersecurity Advisors" 🤔

dav3r replied:
That's good enough for me to change it from three words to two.  I'll chalk it up to the [top-level page](https://www.cisa.gov/about/regions/security-advisors) being outdated. 😄

jsf9k replied:
I was just copying the language from the page that @dav3r linked.  I'm glad @mcdonnnj found evidence for saner orthography.

mcdonnnj commented on cyhy/mailer/cli.py:
```diff
@@ -31,6 +31,27 @@
+  -e --csa-email-file=FILENAME      A YAML file associating each state
+                                    or territory with the email
+                                    address of the corresponding CISA
+                                    Cyber Security Advisor (CSA).  The
```
```suggestion
                                    Cybersecurity Advisor (CSA).  The
```

dav3r replied:
Per https://www.cisa.gov/about/regions/security-advisors, we actually do refer to CSAs as "Cyber Security Advisors" (sigh).

mcdonnnj commented on README.md:
I would really like an example of the CSA emails YAML file in the README (database creds file would be nice too for completeness).

dav3r replied:
Please have a look at https://github.com/cisagov/cyhy-mailer/pull/101/commits/5b7da2cb9ed065aa392a100ed717ec5c28d7c402 and let me know if that works.

mcdonnnj commented on cyhy/mailer/cli.py:
```diff
@@ -697,13 +720,15 @@ def send_cybex_scorecard(
 
-def send_cyhy_reports(db, batch_size, ses_client, cyhy_report_dir, dry_run=False):
+def send_cyhy_reports(
+    db, batch_size, ses_client, cyhy_report_dir, csa_emails, dry_run=False
```
Thoughts?

```suggestion
    db, batch_size, ses_client, cyhy_report_dir, csa_emails=None, dry_run=False
```

dav3r replied:
Good idea- please see https://github.com/cisagov/cyhy-mailer/pull/101/commits/b992f2d8bfe1edef6978055ff9a20e6c9656b12e.

mcdonnnj commented on cyhy/mailer/cli.py:
```diff
@@ -887,14 +946,14 @@ def send_cyhy_reports(db, batch_size, ses_client, cyhy_report_dir, dry_run=False
 
 def send_cyhy_notifications(
-    db, batch_size, ses_client, cyhy_notification_dir, dry_run=False
+    db, batch_size, ses_client, cyhy_notification_dir, csa_emails, dry_run=False
```
```suggestion
    db, batch_size, ses_client, cyhy_notification_dir, csa_emails=None, dry_run=False
```

mcdonnnj commented on README.md:
```diff
@@ -12,26 +12,25 @@ e-mail addresses.
 [--cybex-scorecard-dir=DIRECTORY] [--cyhy-notification-dir=DIRECTORY]
-[--db-creds-file=FILENAME] [--batch-size=SIZE] [--summary-to=EMAILS] [--debug]
-[--dry-run]
+[--db-creds-file=FILENAME] [--csa-email-file=FILENAME] [--batch-size=SIZE]
```
Almost forgot the README.

```suggestion
[--db-creds-file=FILENAME] [--csa-emails-file=FILENAME] [--batch-size=SIZE]
```

mcdonnnj commented on README.md:
```diff
@@ -56,6 +55,26 @@ Options:
   -c --db-creds-file=FILENAME       A YAML file containing the Cyber
                                     Hygiene database credentials.
                                     [default: /run/secrets/database_creds.yml]
+  -e --csa-email-file=FILENAME      A YAML file associating each state or
```
```suggestion
  -e --csa-emails-file=FILENAME     A YAML file associating each state or
```

mcdonnnj commented on cyhy/mailer/cli.py:
```diff
@@ -6,7 +6,7 @@
 
 Usage:
-  cyhy-mailer (bod1801|cybex|cyhy|notification)... [--cyhy-report-dir=DIRECTORY] [--tmail-report-dir=DIRECTORY] [--https-report-dir=DIRECTORY] [--cybex-scorecard-dir=DIRECTORY] [--cyhy-notification-dir=DIRECTORY] [--db-creds-file=FILENAME] [--batch-size=SIZE] [--summary-to=EMAILS] [--debug] [--dry-run]
+  cyhy-mailer (bod1801|cybex|cyhy|notification)... [--cyhy-report-dir=DIRECTORY] [--tmail-report-dir=DIRECTORY] [--https-report-dir=DIRECTORY] [--cybex-scorecard-dir=DIRECTORY] [--cyhy-notification-dir=DIRECTORY] [--db-creds-file=FILENAME] [--csa-email-file=FILENAME] [--batch-size=SIZE] [--summary-to=EMAILS] [--debug] [--dry-run]
```
```suggestion
  cyhy-mailer (bod1801|cybex|cyhy|notification)... [--cyhy-report-dir=DIRECTORY] [--tmail-report-dir=DIRECTORY] [--https-report-dir=DIRECTORY] [--cybex-scorecard-dir=DIRECTORY] [--cyhy-notification-dir=DIRECTORY] [--db-creds-file=FILENAME] [--csa-emails-file=FILENAME] [--batch-size=SIZE] [--summary-to=EMAILS] [--debug] [--dry-run]
```

mcdonnnj commented on cyhy/mailer/cli.py:
```diff
@@ -1060,6 +1153,35 @@ def main():
         return 1
 
+    # Parse out the CSA email mapping, if present.
+    csa_email_file = args["--csa-email-file"]
+    if csa_email_file is not None:
+        try:
+            with open(csa_email_file) as stream:
+                csa_email_mapping = yaml.load(stream, Loader=yaml.SafeLoader)
+        except OSError:
+            logging.critical(
+                f"CSA email file {csa_email_file} does not exist", exc_info=True
+            )
+            return 1
+        except yaml.YAMLError:
+            logging.critical(
+                f"CSA email file {csa_email_file} does not contain valid YAML",
```
```suggestion
    csa_emails_file = args["--csa-emails-file"]
    if csa_emails_file is not None:
        try:
            with open(csa_emails_file) as stream:
                csa_email_mapping = yaml.load(stream, Loader=yaml.SafeLoader)
        except OSError:
            logging.critical(
                f"CSA emails file {csa_emails_file} does not exist", exc_info=True
            )
            return 1
        except yaml.YAMLError:
            logging.critical(
                f"CSA emails file {csa_emails_file} does not contain valid YAML",
```

dav3r replied:
Note @mcdonnnj that I edited your suggestion above to pluralize "emails" in both critical logging messages.

mcdonnnj reviewed (changes requested):
Strong work 💪💪💪 I have some feedback for your consideration.

jsf9k commented:
@dav3r - Please don't merge upstream changes like that if you can help it. It creates an unnecessary merge commit.  Please rebase instead.

dav3r committed:
Pluralize CSA emails file argument name

The file associates each region with a (presumably) different CSA email, so it makes sense to call it `--csa-emails-file` instead of `--csa-email-file`.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Sort projection fields alphabetically

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add additional typing and arguments to mypy pre-commit config

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Set default value None for csa_emails argument

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove an unneeded file

This project uses native namespace packages
(https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages)
so this file is not needed.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add examples of database credentials and CSA emails files to README

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Use more common spelling of "Cybersecurity"

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+---
+- email: csa_region_01@example.gov
+  name: Region 01
+  states_and_territories: [CT, MA, ME, NH, RI, VT]
```
Since we prefer block-style.

```suggestion
  states_and_territories:
    - CT
    - MA
    - ME
    - NH
    - RI
    - VT
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [CT, MA, ME, NH, RI, VT]
+- email: csa_region_02@example.gov
+  name: Region 02
+  states_and_territories: [NJ, NY, PR, VI]
```
```suggestion
  states_and_territories:
    - NJ
    - NY
    - PR
    - VI
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [NJ, NY, PR, VI]
+- email: csa_region_03@example.gov
+  name: Region 03
+  states_and_territories: [DE, DC, MD, PA, VA, WV]
```
```suggestion
  states_and_territories:
    - DE
    - DC
    - MD
    - PA
    - VA
    - WV
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [DE, DC, MD, PA, VA, WV]
+- email: csa_region_04@example.gov
+  name: Region 04
+  states_and_territories: [AL, FL, GA, KY, MS, NC, SC, TN]
```
```suggestion
  states_and_territories:
    - AL
    - FL
    - GA
    - KY
    - MS
    - NC
    - SC
    - TN
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [AL, FL, GA, KY, MS, NC, SC, TN]
+- email: csa_region_05@example.gov
+  name: Region 05
+  states_and_territories: [IL, IN, MI, MN, OH, WI]
```
```suggestion
  states_and_territories:
    - IL
    - IN
    - MI
    - MN
    - OH
    - WI
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [IL, IN, MI, MN, OH, WI]
+- email: csa_region_06@example.gov
+  name: Region 06
+  states_and_territories: [AR, LA, NM, OK, TX]
```
```suggestion
  states_and_territories:
    - AR
    - LA
    - NM
    - OK
    - TX
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [AR, LA, NM, OK, TX]
+- email: csa_region_07@example.gov
+  name: Region 07
+  states_and_territories: [IA, KS, MO, NE]
```
```suggestion
  states_and_territories:
    - IA
    - KS
    - MO
    - NE
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [IA, KS, MO, NE]
+- email: csa_region_08@example.gov
+  name: Region 08
+  states_and_territories: [CO, MT, ND, SD, UT, WY]
```
```suggestion
  states_and_territories:
    - CO
    - MT
    - ND
    - SD
    - UT
    - WY
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [CO, MT, ND, SD, UT, WY]
+- email: csa_region_09@example.gov
+  name: Region 09
+  states_and_territories: [AZ, CA, HI, NV, AS, GU, MP]
```
```suggestion
  states_and_territories:
    - AZ
    - CA
    - HI
    - NV
    - AS
    - GU
    - MP
```

mcdonnnj commented on README.md:
```diff
@@ -69,6 +88,52 @@ Options:
+  states_and_territories: [AZ, CA, HI, NV, AS, GU, MP]
+- email: csa_region_10@example.gov
+  name: Region 10
+  states_and_territories: [AK, ID, OR, WA]
```
```suggestion
  states_and_territories:
    - AK
    - ID
    - OR
    - WA
```

mcdonnnj commented on README.md:
````diff
@@ -69,6 +88,52 @@ Options:
+### Example database credentials file ###
+
+```yaml
+version: '1'
+
+database:
+  name: cyhy
+  uri: mongodb://username:password@hostname:27017/cyhy
````
Just to format as we prefer and add a document start marker.

```suggestion
---
database:
  name: cyhy
  uri: mongodb://username:password@hostname:27017/cyhy
version: '1'
```

mcdonnnj reviewed (commented):
Just some follow up suggestions.

dav3r committed:
Add YAML document start marker and alphabetize keys

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Prefer block-style formatting in YAML example in README

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Bump version from 1.6.0-rc.1 to 1.6.0

dav3r commented:
These changes have been manually deployed to Production - the following files were updated with the changes from this PR and #104:
* `docker-compose.yml`
* `docker-compose.cyhy.yml`
* `docker-compose.cyhy-notification.yml`
* `secrets/csa_emails.yml`
